File: backend/app/api/shopping.py
"""app/api/shopping.py

FastAPI router for shopping list endpoints.
"""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.database.db import get_session
from app.dtos.shopping_dtos import (
    BulkOperationResultDTO,
    BulkStateUpdateDTO,
    ManualItemCreateDTO,
    ShoppingItemResponseDTO,
    ShoppingItemUpdateDTO,
    ShoppingListFilterDTO,
    ShoppingListGenerationDTO,
    ShoppingListGenerationResultDTO,
    ShoppingListResponseDTO,
)
from app.models.user import User
from app.services.shopping_service import ShoppingService

router = APIRouter()


@router.get("", response_model=ShoppingListResponseDTO)
def get_shopping_list(
    source: Optional[str] = Query(None, pattern="^(recipe|manual)$"),
    category: Optional[str] = Query(None),
    have: Optional[bool] = Query(None),
    search_term: Optional[str] = Query(None),
    limit: Optional[int] = Query(None, ge=1, le=100),
    offset: Optional[int] = Query(None, ge=0),
    auto_generate: bool = Query(False, description="DEPRECATED: no longer used, kept for API compatibility"),
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get the shopping list with optional filters.

    Note: auto_generate parameter is deprecated. Shopping list is now automatically
    synced when planner changes occur. This parameter is accepted but ignored.
    """
    service = ShoppingService(session, current_user.id)

    filters = ShoppingListFilterDTO(
        source=source,
        category=category,
        have=have,
        search_term=search_term,
        limit=limit,
        offset=offset,
    )

    return service.get_shopping_list(filters)


@router.get("/items/{item_id}", response_model=ShoppingItemResponseDTO)
def get_shopping_item(
    item_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get a single shopping item by ID."""
    service = ShoppingService(session, current_user.id)
    item = service.shopping_repo.get_shopping_item_by_id(item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Shopping item not found")
    return service._item_to_response_dto(item)


@router.post("/items", response_model=ShoppingItemResponseDTO, status_code=201)
def add_manual_item(
    item_data: ManualItemCreateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Add a manual item to the shopping list."""
    service = ShoppingService(session, current_user.id)
    item = service.add_manual_item(item_data)
    if not item:
        raise HTTPException(status_code=500, detail="Failed to add item")
    return item


@router.patch("/items/{item_id}", response_model=ShoppingItemResponseDTO)
def update_shopping_item(
    item_id: int,
    update_data: ShoppingItemUpdateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Update a shopping item."""
    service = ShoppingService(session, current_user.id)
    item = service.update_item(item_id, update_data)
    if not item:
        raise HTTPException(status_code=404, detail="Shopping item not found")
    return item


@router.patch("/items/{item_id}/toggle", response_model=dict)
def toggle_item_status(
    item_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Toggle the 'have' status of a shopping item."""
    service = ShoppingService(session, current_user.id)
    result = service.toggle_item_status(item_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Shopping item not found")
    return {"success": result}


@router.patch("/items/{item_id}/toggle-flag", response_model=dict)
def toggle_item_flagged(
    item_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Toggle the 'flagged' status of a shopping item."""
    service = ShoppingService(session, current_user.id)
    result = service.toggle_item_flagged(item_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Shopping item not found")
    return {"success": True, "flagged": result}


@router.delete("/items/{item_id}")
def delete_shopping_item(
    item_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Delete a shopping item."""
    service = ShoppingService(session, current_user.id)
    if not service.delete_item(item_id):
        raise HTTPException(status_code=404, detail="Shopping item not found")
    return {"message": "Item deleted successfully"}


@router.post("/generate", response_model=ShoppingListGenerationResultDTO)
def generate_shopping_list(
    generation_data: ShoppingListGenerationDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Generate shopping list from recipe IDs."""
    service = ShoppingService(session, current_user.id)
    return service.generate_shopping_list(generation_data)


@router.post("/generate-from-planner", response_model=ShoppingListGenerationResultDTO)
def generate_from_planner(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """DEPRECATED: Trigger manual shopping list sync.

    Note: This endpoint is deprecated. Shopping list is now automatically synced
    when planner changes occur. This endpoint still works for backwards compatibility
    but is no longer necessary to call.
    """
    service = ShoppingService(session, current_user.id)
    return service.generate_from_active_planner()


@router.delete("/clear", response_model=BulkOperationResultDTO)
def clear_shopping_list(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Clear the entire shopping list."""
    service = ShoppingService(session, current_user.id)
    return service.clear_shopping_list()


@router.delete("/clear-manual", response_model=BulkOperationResultDTO)
def clear_manual_items(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Clear only manual items from the shopping list."""
    service = ShoppingService(session, current_user.id)
    return service.clear_manual_items()


@router.delete("/clear-recipe", response_model=BulkOperationResultDTO)
def clear_recipe_items(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Clear only recipe-generated items from the shopping list."""
    service = ShoppingService(session, current_user.id)
    return service.clear_recipe_items()


@router.delete("/clear-completed", response_model=dict)
def clear_completed_items(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Clear all completed (have=True) items from the shopping list."""
    service = ShoppingService(session, current_user.id)
    count = service.clear_completed_items()
    return {"message": f"Cleared {count} completed items", "deleted_count": count}


@router.patch("/bulk-update", response_model=BulkOperationResultDTO)
def bulk_update_status(
    update_data: BulkStateUpdateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Bulk update 'have' status for multiple items."""
    service = ShoppingService(session, current_user.id)
    return service.bulk_update_status(update_data)


@router.get("/breakdown", response_model=List[dict])
def get_ingredient_breakdown(
    recipe_ids: str = Query(..., description="Comma-separated list of recipe IDs"),
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get ingredient breakdown by recipe."""
    try:
        ids = [int(id.strip()) for id in recipe_ids.split(",") if id.strip()]
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid recipe IDs format")

    service = ShoppingService(session, current_user.id)
    breakdown = service.get_ingredient_breakdown(ids)

    # Convert to list of dicts for response
    result = []
    for item in breakdown.items:
        result.append({
            "ingredient_name": item.ingredient_name,
            "total_quantity": item.total_quantity,
            "unit": item.unit,
            "recipe_contributions": [
                {
                    "recipe_name": rb.recipe_name,
                    "quantity": rb.quantity,
                    "unit": rb.unit,
                    "usage_count": rb.usage_count,
                }
                for rb in item.recipe_breakdown
            ]
        })
    return result

File: frontend/src/lib/api/shopping.ts
import type {
  ShoppingListResponseDTO,
  ShoppingItemResponseDTO,
  ShoppingListGenerationResultDTO,
  ManualItemCreateDTO,
  ShoppingItemUpdateDTO,
  ShoppingListFilterDTO,
  ShoppingListGenerationDTO,
} from "@/types/shopping";
import type { IngredientBreakdownDTO } from "@/types/recipe";
import type { BulkOperationResultDTO } from "@/types/common";
import { fetchApi, buildQueryString } from "./base";

export const shoppingApi = {
  /**
   * Get shopping list with optional filters.
   * Shopping list is automatically synced when planner changes occur.
   * @param filters - Optional filter parameters
   * @param token - Optional auth token for authenticated requests
   */
  getList: (
    filters?: ShoppingListFilterDTO,
    token?: string | null
  ): Promise<ShoppingListResponseDTO> => {
    const query = filters ? buildQueryString(filters as Record<string, unknown>) : "";
    return fetchApi<ShoppingListResponseDTO>(`/api/shopping${query}`, undefined, token);
  },

  /**
   * Get a single shopping item
   * @param id - Shopping item ID
   * @param token - Optional auth token for authenticated requests
   */
  getItem: (id: number, token?: string | null): Promise<ShoppingItemResponseDTO> =>
    fetchApi<ShoppingItemResponseDTO>(`/api/shopping/items/${id}`, undefined, token),

  /**
   * Add a manual item
   * @param data - Manual item data
   * @param token - Optional auth token for authenticated requests
   */
  addItem: (data: ManualItemCreateDTO, token?: string | null): Promise<ShoppingItemResponseDTO> =>
    fetchApi<ShoppingItemResponseDTO>(
      "/api/shopping/items",
      {
        method: "POST",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Update a shopping item
   * @param id - Shopping item ID
   * @param data - Item data to update
   * @param token - Optional auth token for authenticated requests
   */
  updateItem: (
    id: number,
    data: ShoppingItemUpdateDTO,
    token?: string | null
  ): Promise<ShoppingItemResponseDTO> =>
    fetchApi<ShoppingItemResponseDTO>(
      `/api/shopping/items/${id}`,
      {
        method: "PATCH",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Toggle item have status
   * @param id - Shopping item ID
   * @param token - Optional auth token for authenticated requests
   */
  toggleItem: (id: number, token?: string | null): Promise<{ success: boolean }> =>
    fetchApi<{ success: boolean }>(
      `/api/shopping/items/${id}/toggle`,
      {
        method: "PATCH",
      },
      token
    ),

  /**
   * Toggle item flagged status
   * @param id - Shopping item ID
   * @param token - Optional auth token for authenticated requests
   */
  toggleFlagged: (id: number, token?: string | null): Promise<{ success: boolean; flagged: boolean }> =>
    fetchApi<{ success: boolean; flagged: boolean }>(
      `/api/shopping/items/${id}/toggle-flag`,
      {
        method: "PATCH",
      },
      token
    ),

  /**
   * Delete a shopping item
   * @param id - Shopping item ID
   * @param token - Optional auth token for authenticated requests
   */
  deleteItem: (id: number, token?: string | null): Promise<void> =>
    fetchApi<void>(`/api/shopping/items/${id}`, { method: "DELETE" }, token),

  /**
   * Generate shopping list from recipes
   * @param data - Generation parameters
   * @param token - Optional auth token for authenticated requests
   */
  generate: (
    data: ShoppingListGenerationDTO,
    token?: string | null
  ): Promise<ShoppingListGenerationResultDTO> =>
    fetchApi<ShoppingListGenerationResultDTO>(
      "/api/shopping/generate",
      {
        method: "POST",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Generate shopping list from active planner entries.
   * @deprecated Shopping list is now automatically synced when planner changes.
   * This method is kept for backwards compatibility but is no longer necessary.
   * @param token - Optional auth token for authenticated requests
   */
  generateFromPlanner: (token?: string | null): Promise<ShoppingListGenerationResultDTO> =>
    fetchApi<ShoppingListGenerationResultDTO>(
      "/api/shopping/generate-from-planner",
      {
        method: "POST",
      },
      token
    ),

  /**
   * Clear entire shopping list
   * @param token - Optional auth token for authenticated requests
   */
  clear: (token?: string | null): Promise<BulkOperationResultDTO> =>
    fetchApi<BulkOperationResultDTO>(
      "/api/shopping/clear",
      {
        method: "DELETE",
      },
      token
    ),

  /**
   * Clear only manual items
   * @param token - Optional auth token for authenticated requests
   */
  clearManual: (token?: string | null): Promise<BulkOperationResultDTO> =>
    fetchApi<BulkOperationResultDTO>(
      "/api/shopping/clear-manual",
      {
        method: "DELETE",
      },
      token
    ),

  /**
   * Clear completed items
   * @param token - Optional auth token for authenticated requests
   */
  clearCompleted: (token?: string | null): Promise<{ message: string; deleted_count: number }> =>
    fetchApi<{ message: string; deleted_count: number }>(
      "/api/shopping/clear-completed",
      { method: "DELETE" },
      token
    ),

  /**
   * Bulk update item status
   * @param itemUpdates - Map of item IDs to their new have status
   * @param token - Optional auth token for authenticated requests
   */
  bulkUpdate: (
    itemUpdates: Record<number, boolean>,
    token?: string | null
  ): Promise<BulkOperationResultDTO> =>
    fetchApi<BulkOperationResultDTO>(
      "/api/shopping/bulk-update",
      {
        method: "PATCH",
        body: JSON.stringify({ item_updates: itemUpdates }),
      },
      token
    ),

  /**
   * Get ingredient breakdown
   * @param recipeIds - Array of recipe IDs
   * @param token - Optional auth token for authenticated requests
   */
  getBreakdown: (recipeIds: number[], token?: string | null): Promise<IngredientBreakdownDTO[]> =>
    fetchApi<IngredientBreakdownDTO[]>(
      `/api/shopping/breakdown?recipe_ids=${recipeIds.join(",")}`,
      undefined,
      token
    ),
};
